### PR TITLE
Use converted input value if internal currency is FIAT

### LINKED ISSUE #3983: When primary currency is fiat, “Insufficient funds” warning/error message is shown
**Description**
When the primary currency is switched to fiat, “Insufficient funds” warning/error message is shown when trying to send funds out
- The screen then just hangs, and doesn’t let user proceed
- Reproduced consistently on BSC, Avalanche, Goerli (tbd whether this is an exhaustive list of networks)
- Doesn't break the core flow but doesn't allow user to send unless user change that fiat setting
- Not reproduced consistently: the “Next” button is lazy to load, takes >30 secs
- No fund loss
- Workaround
-- App doesn't crash and UI is still intact.  Workaround is to switch back to native in settings
- Hypothesis that needs investigating
-- looks like when you switch to fiat it still looking at the native balance, not the fiat balance

**Screenshots**
- On iOS production as old as v4.0.1: http://recordit.co/0SU7AJctIS
- On Android production as old as v3.1.0: http://recordit.co/TTW0CWCmqt

**To Reproduce**
1. Set primary currency to Fiat in general settings, 
2. navigate to send funds screen, 
3. set any fiat value, 
4. Present with “insufficient funds” warning regardless of balance in account
5. Some team members experienced the "Next" button hanging / not responsive; others didn't experience this. To be investigated

**Expected behavior**
When user switches to fiat, the "Insufficient funds" msg do not show up

**Smartphone (please complete the following information):**
 - OS: iOS, Android

-------------------------------------------------------------
_to be added after bug submission by internal support / PM_
**Severity**
Sev1
https://consensys.slack.com/archives/CBW7S9FSN/p1648234626465419?thread_ts=1648233294.613979&cid=CBW7S9FSN

- How critical is the impact of this bug on a user?
- Add stats if available on % of customers impacted
- Is this visible to all users? Yes
- Is this tech debt?

### PR DESCRIPTION
**Pull request by Fatxx**
**Description**

This PR adds the following changes to Amount screen:

- Inserted amount is now validated correctly when Fiat is selected as primary currency in Settings;
- It should now provide converted Fiat values to all tokens that have conversion rate available, not only for mainnet;
- Improve safety while converting `number` to `BN`;
- Fix inconsistent decimal cases between native tokens and ERC-20 tokens when converted to Fiat;

**What to test**

- ***Case 1 - Insufficient funds fix***
   - Switch to "Binance Smart Chain Mainet"
   - Switch to "Fiat" as primary currency under "Settings > General"
   - Go to Send view
   - Send a value under your balance, it should proceed.

- ***Case 2 - Insufficient funds fix***
   - Switch to "Binance Smart Chain Mainet"
   - Switch to "Fiat" as primary currency under "Settings > General"
   - Go to Send view
   - Send a value above your balance, it should error "Insufficient funds".

- ***Case 3 - Convert FIAT value for tokens (non-ETH)***
   - Switch to "Binance Smart Chain Mainet" (other networks can also be tested)
   - Switch to "Fiat" as primary currency under "Settings > General"
   - Go to send view
   - Insert FIAT value
   - It should show converted BNB value

Note: To validate correct conversion values please compare it with [cryptocompare](https://www.cryptocompare.com/)

***Screenshot***
Screen Shot 2022-12-28 at 16 10 09

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [ ] Any added code is fully documented

**Issue**

Resolves #3983

### CODE REVIEW

File: app/components/Views/SendFlow/Amount/index.js [hunk 1]
```diff
@@ -638,14 +637,18 @@ class Amount extends PureComponent {
...
 	 */
-	validateAmount = (inputValue) => {
-		const { accounts, selectedAddress, contractBalances, selectedAsset } = this.props;
-		const { estimatedTotalGas } = this.state;
+	validateAmount = (inputValue, internalPrimaryCurrencyIsCrypto) => {
+		const { accounts, selectedAddress, selectedAsset, contractBalances } = this.props;
+		const { estimatedTotalGas, inputValueConversion } = this.state;
+
+		if (!internalPrimaryCurrencyIsCrypto) {
+			inputValue = inputValueConversion;
```
Comment (jpcloureiro) [inline app/components/Views/SendFlow/Amount/index.js:649]: I suggest creating a new variable instead of mutating the function parameter `inputValue`

**Review by CryptoGodfatherVA4 — APPROVED** [review]

**Review by CryptoGodfatherVA4 — APPROVED** [review]

**Review by JessePinkman48 — APPROVED** [review]

**Comment by chrisleewilcox** [comment]
This branch is a draft.  Please update PR to show green open icon.  Makes it easy to determine if a PR is ready for QA.

**Review by Fatxx — COMMENTED** [review]
What do to when we don't have conversion rate available? 

- Force the user to switch to crypto as native currency?

@jakehaugen @yanrong-chen

**Comment by jakehaugen** [comment]
> What do to when we don't have conversion rate available?
> 
> * Force the user to switch to crypto as native currency?

Discussed with @Fatxx, will work on some mocks and follow up.

**Comment by Fatxx** [comment]
Figma: https://www.figma.com/file/ad2Higmy1dBMonznFEV4hZ/Fiat-converssions-not-available?node-id=5%3A102

File: app/components/Views/SendFlow/Amount/index.js [hunk 2]
```diff
@@ -485,15 +493,40 @@ class Amount extends PureComponent {
...
     }
 
+    const hasExchangeRate = this.hasExchangeRate();
+    let internalPrimaryCurrencyIsCrypto =
+      this.state.internalPrimaryCurrencyIsCrypto;
+
+    // Default to crypto if exchange rate is not available while on Fiat primary currency
+    if (this.props.primaryCurrency === 'Fiat' && !hasExchangeRate) {
+      internalPrimaryCurrencyIsCrypto = true;
+    }
```
[the remark this answers is not in the document]
Comment (Fatxx) [reply app/components/Views/SendFlow/Amount/index.js:503]: If it's crypto it uses the same logic that we have before

**Review by gantunesr — APPROVED** [review]
LGTM

Also verifies the correct functionality of sending ERC721 tokens and the same use cases as before in Ethereum Mainnet.

**Comment by chrisleewilcox** [comment]
@Fatxx can you resolve the merge conflict?

**Comment by cortisiko** [comment]
@Fatxx can you merge in main and solve these merge conflicts? I will remove needs QA for now. cc @sethkfman

**Review by cortisiko — CHANGES_REQUESTED** [review]
@Fatxx swinging this back your way:

**Issue 1**
is [this](https://www.screencast.com/t/eq4dBRn4GGT) expected? Happens when I attempt to send an NFT. 

**Issue 2**
if my primary currency is set to Fiat and I send funds, when I view the transaction activity, [the default seems to be Native and not Fiat](https://www.screencast.com/t/4Qm1tzltyB).

**Issue 3**
My primary currency is set to Fiat, when I view the detected tokens in my wallet, [the default currency is native and not Fiat. ](https://www.screencast.com/t/x8EmWJ20h)

**Issue 4**
if I were to scan a payment request QR code, when I land on the transaction confirmation view, the fiat conversion is set to 0. See [recording](https://recordit.co/MFv5VheCHm)

To reproduce:
on device A, request funds (go through the receive flow) then open up the QR code for the payment request
on device B scan device A QR code. Notice the fiat value is set to 0 

**Issue 5**
With Fiat as my default currency, when i tap on the send button on the transaction confirmation screen, the screen does not disappear. See [recording](http://recordit.co/nvue2E2eE7).



To reproduce:
add then switch to BSC
change your default currency from Native to Fiat
attempt to send funds, notice after you tap on “send” the transaction confirmation view does not disappear.

**Comment by Fatxx** [comment]
@cortisiko 

I fixed **Issue 1** since it's related to the Amount Screen, which was the initial screen that needed to be set in this PR.

I'll look at **4** and **5** since they are related to the overall transaction flow.

Issues **2** and **3** are out of scope and should be developed in separate issues.

**Comment by Fatxx** [comment]
@cortisiko 

Regarding **issue 4**:
The problem here is that the sender doesn't have the token currently on the token list and the controllers fetch the exchange rates from this token list, so when you open the confirm screen from the QR code the conversion rate is not available.

What I did to fix this is to add the respective erc20 token to the token list, since there is no way to get this exchange rate directly from the controllers. cc @sethkfman 

Recording with the fix:
https://recordit.co/vK5sfaVBoc

**Comment by Fatxx** [comment]
@cortisiko 

Regarding **issue 5**, I'm not able to reproduce it.

https://recordit.co/pXeQTh8SUY

**Comment by sethkfman** [comment]
Test case
- User receiving payment request
   - User has 0 token balance (temporarily add during send flow, don't show token)
   - User has more than 0 token balance but it doesn't display in wallet (add token and display)

**Comment by Fatxx** [comment]
@cortisiko 

**Issue 4** fixed

**Comment by cortisiko** [comment]
When I send funds, I get `[TypeError: undefined is not an object (evaluating 'weiBalance.isZero')`


to reproduce:
send funds to an address
notice the error. 

@Fatxx sending this back your way.

**Review by cortisiko — APPROVED** [review]
ticket looks good to me 🌮 🌮 🌮

**Pull request closed without merging**